### PR TITLE
Add select component

### PR DESCRIPTION
**Pull request by Ufooo**

### CODE REVIEW

**Comment by driesvints** [comment]
I don't have time to look at the PR atm but hopefully soon. In the meantime can you rebase and solve the merge conflicts?

**Comment by Ufooo** [comment]
it's done

File: config/blade-ui-kit.php [hunk 1]
```diff
@@ -31,6 +31,10 @@
         'form-button' => Components\Buttons\FormButton::class,
         'html' => Components\Layouts\Html::class,
         'input' => Components\Forms\Inputs\Input::class,
+        'select' => Components\Forms\Inputs\Select::class,
+        'checkbox' => Components\Forms\Inputs\Checkbox::class,
+        'color-picker' => Components\Forms\Inputs\ColorPicker::class,
+        'email' => Components\Forms\Inputs\Email::class,
```
Comment (driesvints) [inline config/blade-ui-kit.php:37]: Seems like some things went wrong and there are now duplicate components here.

**Review by Ufooo — COMMENTED** [review]
It's done

File: config/blade-ui-kit.php [hunk 2]
```diff
@@ -31,6 +31,7 @@
         'form-button' => Components\Buttons\FormButton::class,
         'html' => Components\Layouts\Html::class,
         'input' => Components\Forms\Inputs\Input::class,
+        'select' => Components\Forms\Inputs\Select::class,
```
Comment (driesvints) [inline config/blade-ui-kit.php:34]: Please alphabetically arrange this.

**Review by Ufooo — COMMENTED** [review]
Fixed

**Comment by freekmurze** [comment]
TL;DR: Let's optimise this component for elequent collections instead of arrays, since you'll likely work much more with eloquent arrays in a typical app.

Right now, this component accepts simple arrays. The keys will be used as values in the html, the values in the array will be used as label in the html.

Imagine you want to create a select element with let's say all departments in your app. Departments are stored in the db. In your controller you might retrieve them like this.

```php
Department::all()
```

In order to pass the models to the view component you'd have to perform a `map` operation.

```php
$departments = Department::all()->mapWithKey(function(Department $department) {
    return [$department->id => $department->name]
});
```

Having to do this will grow old pretty fast.

I propose to change this component so that it accepts  an `iterable` instead of an only array, so a `Collection` gets accepted as well. By default the `id` attribute could be used as value in the html and the `name` attribute as the label in the html.

This way you don't need to do any transformation and this would just work.

```php
<x-select :values="$departments" />
```

Optionally, `value-attribute` and `name-attribute` could be accepted to override the `id` and `name` defaults.

This is just a rough sketch, I hope you get the idea. Alternatively a second Blade component could be introduced (could be named `x-select- collection`.

**Comment by LonnyX** [comment]
@driesvints can we merge this? It sounds like a miss on the package :/

**Comment by driesvints** [comment]
@LonnyX sorry I haven't gotten to this yet. I've been a bit swamped the last couple of weeks so I'm running behind on issues/prs on Blade UI Kit. I'm hoping to get through them somewhere soon although it's hard to say when exactly.

I also want to think out this particularly PR a bit better before merging like Freek suggests.

**Comment by nasrulhazim** [comment]
@Ufooo , if i make it as multiple selection, i won't be able to set selected values.

**Comment by Ufooo** [comment]
@freekmurze did you think so? :)

**Comment by freekmurze** [comment]
@Ufooo Nice!

**Comment by riseno** [comment]
when will this PR be merged?

**Comment by danharrin** [comment]
I think we need to make some edits, @Ufooo are you willing to still maintain this PR or would you like me to fork and make the changes on my own?

File: resources/views/components/forms/inputs/select.blade.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<select
+    name="{{ $name }}"
+    id="{{ $id }}"
+    {{ $attributes }}
+>
+@if(!$slot->isEmpty())
```
Comment (danharrin) [inline resources/views/components/forms/inputs/select.blade.php:6]:
```suggestion
@if(! $slot->isEmpty())
```

File: src/Components/Forms/Inputs/Select.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        $this->name = $name;
+        $this->id = $id ?? $name;
+        $this->nameAttribute = $nameAttribute;
+        $this->valueAttribute = $valueAttribute;
+        $this->selected = $selected;
+        $this->placeholder = $placeholder;
+
+        if ($options instanceof Collection) {
+            $this->options = $options->mapWithKeys(function ($option) {
```
Comment (danharrin) [inline src/Components/Forms/Inputs/Select.php:45]: We can refactor this `mapWithKeys()` expression to use `pluck()`

```suggestion
            $this->options = $options->pluck($this->nameAttribute, $this->valueAttribute)
```

File: src/Components/Forms/Inputs/Select.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function isSelected($value): bool
+    {
+        $selected = old($this->name, $this->selected);
+
+        if (is_array($selected)) {
+            return in_array($value, $selected);
+        }
+
+        return ($value === $selected);
```
Comment (danharrin) [inline src/Components/Forms/Inputs/Select.php:63]:
```suggestion
        return $value === $selected;
```

File: src/Components/Forms/Inputs/Select.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        } elseif (is_array($options)) {
+            $this->options = $options;
+        } else {
+            throw new InvalidArgumentException('Invalid options');
+        }
+    }
+
+    public function isSelected($value): bool
+    {
+        $selected = old($this->name, $this->selected);
```
Comment (danharrin) [inline src/Components/Forms/Inputs/Select.php:57]: What is the purpose of the `old()` call here?

File: src/Components/Forms/Inputs/Select.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    /** @var string */
+    public $valueAttribute;
+
+    /** @var string|array|null */
+    public $selected;
+
+    /** @var string */
+    public $placeholder;
+
+    public function __construct(string $name, $options, string $placeholder = '', string $nameAttribute = 'name', string $valueAttribute = 'id', $id = null, $selected = null)
```
Comment (danharrin) [inline src/Components/Forms/Inputs/Select.php:35]: I think that a more appropriate name for `$nameAttribute` is probably `$labelAttribute` - each option has a (display) label and value. Thoughts?

**Review by danharrin — CHANGES_REQUESTED** [review]

**Comment by howdu** [comment]
Until this gets merged [laravel-form-components](https://github.com/rawilk/laravel-form-components) has a complete selection of form components.

**Comment by riseno** [comment]
> when will this PR be merged?

well, am gonna leave a footprint for this PR almost 20 months after.

**Comment by iglebekk** [comment]
I would be nice to add this to the kit.
